### PR TITLE
Routing cleanup

### PR DESCRIPTION
**Pull request by alisa911**

### CODE REVIEW

File: native/src/binaryRoutePlanner.cpp [hunk 1]
```diff
@@ -455,115 +451,109 @@ bool checkIfOppositieSegmentWasVisited(RoutingContext* ctx, bool reverseWaySearc
...
+		if (existingSegment != nullptr) {
+			if (distFromStartPlusSegmentTime > existingSegment->distanceFromStart) {
+                // insert back original segment (test case with large area way)
+				directionAllowed = false;
+				if (TRACE_ROUTING) {
+                    OsmAnd::LogPrintf(OsmAnd::LogSeverityLevel::Info, "  >> Already visited");
+				}
+				break;
+			} else {
+                visitedSegments[nextPntId] = currentSegment;
```
Comment (vshcherb) [inline native/src/binaryRoutePlanner.cpp:540]: formatting

File: native/src/binaryRoutePlanner.cpp [hunk 2]
```diff
@@ -142,10 +142,14 @@ SHARED_PTR<RouteSegment> initRouteSegment(RoutingContext* ctx, SHARED_PTR<RouteS
...
 							SEGMENTS_QUEUE& graphDirectSegments, SEGMENTS_QUEUE& graphReverseSegments) {
-	SHARED_PTR<RouteSegment> startPos = initRouteSegment(ctx, start, true);
-	SHARED_PTR<RouteSegment> startNeg = initRouteSegment(ctx, start, false);
-	SHARED_PTR<RouteSegment> endPos = initRouteSegment(ctx, end, true);
-	SHARED_PTR<RouteSegment> endNeg = initRouteSegment(ctx, end, false);
+	SHARED_PTR<RouteSegment> startPos = initRouteSegment(ctx, start, true, false);
+	SHARED_PTR<RouteSegment> startNeg = initRouteSegment(ctx, start, false, false);
+	SHARED_PTR<RouteSegment> endPos = initRouteSegment(ctx, end, true, true);
+	SHARED_PTR<RouteSegment> endNeg = initRouteSegment(ctx, end, false, true);
+    startPos->parentRoute = routeSegmentNULL;
```
Comment (vshcherb) [inline native/src/binaryRoutePlanner.cpp:149]: formatting

File: native/src/routingContext.h [hunk 3]
```diff
@@ -468,11 +471,21 @@ struct RoutingContext {
 					SHARED_PTR<RouteDataObject> toCmp = excludeDuplications[calcRouteId(ro, segment->getSegmentStart())];
 					if (!isExcluded(ro->id, j, subregions) && (toCmp.get() == NULL || toCmp->pointsX.size() < ro->pointsX.size())) {
 						excludeDuplications[calcRouteId(ro, segment->getSegmentStart())] = ro;
-						SHARED_PTR<RouteSegment> s = std::make_shared<RouteSegment>(ro, segment->getSegmentStart());
-						s->next = original;
-						original = 	s;
+//						SHARED_PTR<RouteSegment> s = std::make_shared<RouteSegment>(ro, segment->getSegmentStart());
+//						s->next = original;
+//						original = 	s;
+                        if (reverseWaySearch) {
```
Comment (vshcherb) [inline native/src/routingContext.h:477]: formatting

File: native/src/routeSegment.h [hunk 4]
```diff
@@ -40,38 +43,52 @@ struct RouteSegment {
...
     
+    bool isSegmentAttachedToStart() {
+        auto parentRoutePtr = parentRoute.lock();
+        if (parentRoutePtr != nullptr) {
+            return true;
+        }
+        return false;
+    }
+    
+    inline std::weak_ptr<RouteSegment> getParentRoute();
```
Comment (vshcherb) [inline native/src/routeSegment.h:67]: parent route can't be weak

File: native/src/routeSegment.h [hunk 5]
```diff
@@ -6,32 +6,35 @@
...
-    SHARED_PTR<RouteSegment>  oppositeDirection ;
+    SHARED_PTR<RouteSegment> nextLoaded;
+    SHARED_PTR<RouteSegment> oppositeDirection;
+    SHARED_PTR<RouteSegment> reverseSearch;
     
     // search context (needed for searching route)
     // Initially it should be null (!) because it checks was it segment visited before
     std::weak_ptr<RouteSegment> parentRoute;
-    uint16_t parentSegmentEnd;
+    //uint16_t parentSegmentEnd;
```
Comment (vshcherb) [inline native/src/routeSegment.h:22]: no comment code is needed

File: native/src/routeSegment.h [hunk 6]
```diff
@@ -6,32 +6,35 @@
...
     // needed to store intersection of routes
     SHARED_PTR<RouteSegment> next;
-    SHARED_PTR<RouteSegment>  oppositeDirection ;
+    SHARED_PTR<RouteSegment> nextLoaded;
+    SHARED_PTR<RouteSegment> oppositeDirection;
+    SHARED_PTR<RouteSegment> reverseSearch;
     
     // search context (needed for searching route)
     // Initially it should be null (!) because it checks was it segment visited before
     std::weak_ptr<RouteSegment> parentRoute;
```
Comment (vshcherb) [inline native/src/routeSegment.h:21]: Weak ptr can't be used?

File: native/src/routeSegment.h [hunk 7]
```diff
@@ -80,29 +97,54 @@ struct RouteSegment {
...
         , reverseWaySearch(0)
         , opposite()
         , distanceFromStart(0)
-        , distanceToEnd(0) {
+        , distanceToEnd(0)
+        , isFinalSegment(false) {
     }
 };
 
+struct RouteSegmentNULL : RouteSegment {
```
Comment (vshcherb) [inline native/src/routeSegment.h:136]: class & constant is not needed

File: native/src/routingContext.h [hunk 8]
```diff
@@ -468,11 +471,18 @@ struct RoutingContext {
 					SHARED_PTR<RouteDataObject> toCmp = excludeDuplications[calcRouteId(ro, segment->getSegmentStart())];
 					if (!isExcluded(ro->id, j, subregions) && (toCmp.get() == NULL || toCmp->pointsX.size() < ro->pointsX.size())) {
 						excludeDuplications[calcRouteId(ro, segment->getSegmentStart())] = ro;
-						SHARED_PTR<RouteSegment> s = std::make_shared<RouteSegment>(ro, segment->getSegmentStart());
-						s->next = original;
-						original = 	s;
+						if (reverseWaySearch) {
+                            if (segment->reverseSearch == NULL) {
```
Comment (vshcherb) [inline native/src/routingContext.h:475]: formatting

File: native/src/java_wrap.cpp [hunk 9]
```diff
@@ -773,6 +777,7 @@ void loadJniRenderingContext(JNIEnv* env) {
 	jclass_TurnType = findGlobalClass(env, "net/osmand/router/TurnType");
 	jmethod_TurnType_init = env->GetMethodID(jclass_TurnType, "<init>", "(IIFZ[IZZ)V");
 
+    jclass_RouteCalculationMode = findGlobalClass(env, "net/osmand/router/RoutePlannerFrontEnd$RouteCalculationMode");
```
Comment (vshcherb) [inline native/src/java_wrap.cpp:780]: formatting

File: native/src/java_wrap.cpp [hunk 10]
```diff
@@ -1575,6 +1583,11 @@ RoutingContext* getRoutingContext(JNIEnv* ienv, jobject jCtx, jfloat initDirecti
 	c->targetRoadId = ienv->GetLongField(jCtx, jfield_RoutingContext_targetRoadId);
 	c->targetSegmentInd = ienv->GetIntField(jCtx, jfield_RoutingContext_targetSegmentInd);
 	c->targetTransportStop = ienv->GetBooleanField(jCtx, jfield_RoutingContext_targetTransportStop);
+    jmethodID getOrdinalValueMethod = ienv->GetMethodID(jclass_RouteCalculationMode, "ordinal", "()I");
+    jobject calculationMode = ienv->GetObjectField(jCtx, jfield_RoutingContext_calculationMode);
```
Comment (vshcherb) [inline native/src/java_wrap.cpp:1587]: formatting

File: native/src/binaryRoutePlanner.cpp [hunk 11]
```diff
@@ -455,119 +459,116 @@ bool checkIfOppositieSegmentWasVisited(RoutingContext* ctx, bool reverseWaySearc
...
-		if (heightObstacle < 0) {
-			directionAllowed = false;
-			break;
+		// calculate new start segment time as we're going to assign to put to visited segments
+		float distFromStartPlusSegmentTime = currentSegment->distanceFromStart + segmentAndObstaclesTime;
+		if (ctx->precalcRoute != nullptr && ctx->precalcRoute->followNext) {
+			// speed up calculation with calculated route by using different distance from start
+			const int x = road->pointsX[currentSegment->getSegmentStart()];
+			const int y = road->pointsY[currentSegment->getSegmentStart()];
+			// TODO double check is it correct (looks OK)
```
Comment (vshcherb) [inline native/src/binaryRoutePlanner.cpp:522]: delete todo

File: native/src/binaryRoutePlanner.cpp [hunk 12]
```diff
@@ -1,6 +1,7 @@
 #include "binaryRoutePlanner.h"
 
 #include <functional>
+#include <iostream>
```
Comment (ivanPyrohivskyi) [inline native/src/binaryRoutePlanner.cpp:4]: Remove it

File: native/src/binaryRoutePlanner.cpp [hunk 13]
```diff
@@ -273,7 +281,7 @@ SHARED_PTR<RouteSegment> searchRouteInternal(RoutingContext* ctx, SHARED_PTR<Rou
 		if (TRACE_ROUTING) {
 			printRoad(forwardSearch ? "F>" : "B>", segment);
 		}
-		if (segment->isFinal()) {
+		if (segment->isFinalSegment == true) {
```
Comment (ivanPyrohivskyi) [inline native/src/binaryRoutePlanner.cpp:284]: Just segment->isFinalSegment

File: native/src/binaryRoutePlanner.cpp [hunk 14]
```diff
@@ -676,90 +670,121 @@ bool proccessRestrictions(RoutingContext* ctx, SHARED_PTR<RouteSegment>& segment
...
+	// reassign @distanceToEnd to make it correct for visited segment
+	currentSegment->distanceToEnd = distanceToEnd;
+
+	SHARED_PTR<RouteSegment> connectedNextSegment = ctx->loadRouteSegment(x, y, reverseWaySearch);
+	SHARED_PTR<RouteSegment> roadIter = connectedNextSegment;
+	bool directionAllowed = true;
+	bool singleRoad = true;
+	while (roadIter != nullptr) {
+		if (currentSegment->getSegmentEnd() == roadIter->getSegmentStart() &&
+			roadIter->road->getId() == currentSegment->getRoad()->getId()) {
```
Comment (ivanPyrohivskyi) [inline native/src/binaryRoutePlanner.cpp:703]: roadIter->road can be nullptr

Comment (ivanPyrohivskyi) [inline native/src/binaryRoutePlanner.cpp:703]: Need same style here, please use getRoad()

File: native/src/binaryRoutePlanner.cpp [hunk 15]
```diff
@@ -919,96 +944,104 @@ void attachConnectedRoads(RoutingContext* ctx, vector<SHARED_PTR<RouteSegmentRes
...
 	if (next.get() != NULL) {
-		double obstaclesTime = ctx->config->router->calculateTurnTime(
-			next,
-			// next->getSegmentStart(),
-			next->isPositive() ? next->road->getPointsLength() - 1 : 0, segment, segmentPoint);
-		distFromStart += obstaclesTime;
+		if (!checkMovementAllowed(ctx, reverseWaySearch, next)) {
+			return false;
+		}
+		float obstaclesTime = (float)ctx->config->router->calculateTurnTime(
```
Comment (ivanPyrohivskyi) [inline native/src/binaryRoutePlanner.cpp:954]: Can be lose 1 during cast double to float. So need to be sure that you do not use so small values here

Comment (ivanPyrohivskyi) [reply native/src/binaryRoutePlanner.cpp:954]: Please add comment here: "//Can lose 1 during cast double to float"

File: native/src/binaryRoutePlanner.cpp (hunk 14 above)
Comment (ivanPyrohivskyi) [reply native/src/binaryRoutePlanner.cpp:703]: Better add check roadIter->getRoad() != nullptr
I am not sure that getRoad() can not be nullptr

Comment (alisa911) [reply native/src/binaryRoutePlanner.cpp:703]: Only start and end segments can have RouteSegment with road == nullptr

**Pull request merged**